test/model/recent_senders_test.dart

```diff
@@ -0,0 +1,329 @@
+  final Map<int, Map<String, Map<int, Set<int>>>> messagesByUserInTopic = {};
+  for (final message in messages) {
+    if (message is! StreamMessage) {
+      throw UnsupportedError('Message of type ${message.runtimeType} is not expected.');
+    }
+
+    final StreamMessage(:streamId, :topic, :senderId, id: int messageId) = message;
+
+    ((messagesByUserInStream[streamId] ??= {})
+      [senderId] ??= {}).add(messageId);
+    (((messagesByUserInTopic[streamId] ??= {})[topic] ??= {})
+      [senderId] ??= {}).add(messageId);
+  }
+
+  final actualMessagesByUserInStream = model.streamSenders.map((streamId, sendersByStream) =>
+    MapEntry(streamId, sendersByStream.map((senderId, tracker) =>
+      MapEntry(senderId, Set<int>.from(tracker.ids)))));
+  final actualMessagesByUserInTopic = model.topicSenders.map((streamId, topicsByStream) =>
+    MapEntry(streamId, topicsByStream.map((topic, sendersByTopic) =>
+      MapEntry(topic, sendersByTopic.map((senderId, tracker) =>
+        MapEntry(senderId, Set<int>.from(tracker.ids)))))));
+
+  check(actualMessagesByUserInStream).deepEquals(messagesByUserInStream);
+  check(actualMessagesByUserInTopic).deepEquals(messagesByUserInTopic);
+}
+
+void main() {
+  test('starts with empty stream and topic senders', () {
+    final model = RecentSenders();
+    checkMatchesMessages(model, []);
+  });
+
+  group('RecentSenders.handleMessage', () {
+    test('stream message gets included', () {
+      final model = RecentSenders();
+      final streamMessage = eg.streamMessage();
+      model.handleMessage(streamMessage);
+      checkMatchesMessages(model, [streamMessage]);
+    });
+
+    test('DM message gets ignored', () {
+      final model = RecentSenders();
+      final dmMessage = eg.dmMessage(from: eg.selfUser, to: [eg.otherUser]);
+      model.handleMessage(dmMessage);
+      checkMatchesMessages(model, []);
+    });
+  });
+
+  group('RecentSenders.handleMessages', () {
+    late RecentSenders model;
+    final stream1 = eg.stream(streamId: 1);
+    final user10 = eg.user(userId: 10);
+
+    void setupModel(List<Message> messages) {
+      model = RecentSenders();
+      for (final message in messages) {
+        model.handleMessage(message);
+      }
+    }
+
+    group('single tracker', () {
+      test('batch goes before the existing messages', () {
+        final existingMessages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 300),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 400),
+        ];
+        setupModel(existingMessages);
+
+        final messages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 100),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 200),
+        ];
+        model.handleMessages(messages);
+
+        checkMatchesMessages(model,
+          [...existingMessages, ...messages]..sort((m1, m2) => m1.id.compareTo(m2.id)));
```

Comment:
These handleMessages tests are pretty verbose, with a lot of stuff in them that's the same from test to test and isn't what the test is about. I'll push a couple of added commits that factor that out, so that it's a lot easier for the reader to scan these tests and see what they're saying.